agent/consul/internal_endpoint.go

```diff
@@ -69,18 +76,67 @@ func (m *Internal) NodeDump(args *structs.DCSpecificRequest,
 		&args.QueryOptions,
 		&reply.QueryMeta,
 		func(ws memdb.WatchSet, state *state.Store) error {
-			index, dump, err := state.NodeDump(ws, &args.EnterpriseMeta, args.PeerName)
-			if err != nil {
-				return err
+			peerings := make([]string, 0)
+			peerings = append(peerings, "") // this represents a local fetch
+
+			// if a PeerName is passed, then it's we only get the NodeDump for that peer
+			// otherwise, we get all the local nodes and all the nodes for every peering
+			if args.PeerName != "" {
+				peerings = append(peerings, args.PeerName)
+			} else {
+				// get a list of all peerings
+				_, listedPeerings, err := state.PeeringList(ws, args.EnterpriseMeta)
+				if err != nil {
+					m.logger.Error("could not list peers for node dump", err)
+				} else {
+					for _, lp := range listedPeerings {
+						peerings = append(peerings, lp.Name)
+					}
+				}
 			}
-			reply.Index, reply.Dump = index, dump
+
+			// maxIndex will be set to the highest index of the NodeDump below
+			var maxIndex uint64
+			for _, p := range peerings {
+				if p != "" {
+					index, importedDump, err := state.NodeDump(ws, &args.EnterpriseMeta, p)
+					if err != nil {
+						return err
+					}
+					reply.ImportedDump = append(reply.ImportedDump, importedDump...)
+
+					if index > maxIndex {
+						maxIndex = index
+					}
+				} else {
+					index, dump, err := state.NodeDump(ws, &args.EnterpriseMeta, "")
+					if err != nil {
+						return err
+					}
+
+					// TODO(peering): we should not ever be here twice but maybe it would be important
+					// to log if we are as that could be a programming error
+					reply.Dump = dump
+
+					if index > maxIndex {
+						maxIndex = index
+					}
+				}
+			}
+			reply.Index = maxIndex
 
 			raw, err := filter.Execute(reply.Dump)
 			if err != nil {
 				return err
 			}
 			reply.Dump = raw.(structs.NodeDump)
 
+			importedRaw, err := filter.Execute(reply.ImportedDump)
```

Comment:
Can we test specifying a filter selector that points at `ImportedDump`?

Reply:
By the way, we added tests for these!